arcs/src/systems/spatial_relation.rs

```diff
@@ -0,0 +1,204 @@
+    use piet::Color;
+
+    #[test]
+    fn setup_creates_all_outstanding_spatial_entities() {
+        let mut world = World::new();
+
+        // make sure we register all components
+        register(&mut world);
+    
+        let layer = Layer::create(
+            world.create_entity(),
+            Name::new("default"),
+            Layer {
+                z_level: 0,
+                visible: true,
+            },
+        );
+    
+        // Add a line to our world
+        let line = Line::new(Vector::new(2.0, 1.0), Vector::new(5.0, -1.0));
+        let _first = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Setup our spatial system
+        let mut system = SpatialRelation::new(&world);
+        System::setup(&mut system, &mut world);
+
+        let space = world.read_resource::<Space>();
+        assert_eq!(space.len(), 1);
+
+    }
+
+    #[test]
+    fn run_will_keep_spatial_updated() {
+        let mut world = World::new();
+
+        // make sure we register all components
+        register(&mut world);
+    
+        let layer = Layer::create(
+            world.create_entity(),
+            Name::new("default"),
+            Layer {
+                z_level: 0,
+                visible: true,
+            },
+        );
+    
+        // Add a line to our world
+        let line = Line::new(Vector::new(2.0, 1.0), Vector::new(5.0, -1.0));
+        let _first = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Setup our spatial system
+        let mut system = SpatialRelation::new(&world);
+        System::setup(&mut system, &mut world);
+        
+        // make some changes after the initial setup
+        let line = Line::new(Vector::new(3.0, 0.0), Vector::new(-1.0, 2.0));
+        let _second = world
+            .create_entity()
+            .with(DrawingObject {
+                geometry: Geometry::Line(line),
+                layer,
+            })
+            .with(LineStyle {
+                width: Dimension::DrawingUnits(5.0),
+                stroke: Color::rgb8(0xff, 0, 0),
+            })
+            .build()
+        ;
+
+        // Test if the system works
+        system.run_now(&world);
+        let space = world.read_resource::<Space>();
+
+        let query = space.query_point(Vector::new(3.0, -0.5));
+        assert!(query != None);
+        assert_eq!(query.unwrap().len(), 1);
```

Comment:
Hmm... After seeing how a typical caller will be using the query, does it actually make sense to return an `Option<Vec<_>>`? 

If we return an empty `Vec` (note: it's guaranteed that empty vecs don't allocate) when the query has no results, the `query_*()` methods should be a lot more ergonomic.

Also, semantically returning an empty `Vec` means your query completed but nothing was found. Which kinda makes sense?

Reply:
True that makes more sense!

Reply:
Done